Document/0x05g-Testing-Network-Communication.md

```diff
@@ -196,3 +196,180 @@ The SSL pinning process should be implemented as described on the static analysi
 * [1] Setting Burp Suite as a proxy for Android Devices -  https://support.portswigger.net/customer/portal/articles/1841101-configuring-an-android-device-to-work-with-burp)
 * [2] OWASP Certificate Pinning for Android - https://www.owasp.org/index.php/Certificate_and_Public_Key_Pinning#Android
 * [3] okhttp library - https://github.com/square/okhttp/wiki/HTTPS
+
+
+### Testing used connectivity and communication libraries
+
+#### Overview 
+Android relies on a security provider to provide SSL/TLS based connections. The problem with this security provider (for instance OpenSSL) which is packed with the device, is that it often has bugs and/or vulnerabilities<sup>[1]</sup>.
+Developers need to make sure that the application will install a proper security provider to make sure that there will be lesser bugs and vulnerabilities.
+
+#### Static Analysis
+In case of an Android SDK based application. The application should have a dependency on the GooglePlayServices. (e.g. in a. gradle build file, you will find `compile 'com.google.android.gms:play-services-gcm:x.x.x'` in the dependencies block). Next you need to make sure that the `ProviderInstaller` class is called with either `installIfNeeded()` or with `installIfNeededAsync()` is called as soon as possible and that errors comming from that are handled correctly: either the application informs the API on his lesser secure state or it restricts the user in its possible actions as all https-traffic should now be deemed more risky.
+
+--- {TODO: ADD ANDROID NATIVE STATIC ANALYSIS}
+
+#### Dynamic Analysis
+When you have the source-code: 
+
+- Run the application in debugmode, then make a debug-point right at the point where the app will make it first contact with the backend.
+- rightclick at the code that is highlighted and select `Evaluate Expression`
+- Type `Security.getProviders()` and press enter
+- Check the providers and see if you can find `GmsCore_OpenSSL` which should be the new toplisted provider.
+
+When you do not have the source-code:
+- Use Xposed to hook into `java.security` package, then hook into `java.security.Security` with the method `getProviders` with no arguments. The return value is an Array of `Provider`. 
+- Check if the first provider is `GmsCore_OpenSSL`.
+
+--- {TODO: ADD ANDROID NATIVE DYNAMIC ANALYSIS}
```

Comment:
Per IRL discussion, perhaps leave the discussion at a high level. Add a reference to the chapter that discusses general secure networking principles. There may be high variability in native/NDK libraries.